src/python/zquantum/core/hamiltonian.py

```diff
@@ -69,3 +71,124 @@ def group_comeasureable_terms_greedy(
+
+    expectations = np.real(np.array(list(expectations_packed.terms.values()))) # should we add an assert to catch large Im parts
+    # Clip expectations if they fell out of [-1 , 1] due to numerical errors
+    expectations[expectations < -1.0] = -1.0
+    expectations[expectations >  1.0] =  1.0
+
+    return ExpectationValues(expectations)
+
+def estimate_nmeas(
+    target_operator: QubitOperator,
+    decomposition_method: Optional[str] = "greedy-sorted",
+    expecval: Optional[ExpectationValues] = None,
+) -> Tuple[float, int, np.array]:
+    """Calculates the number of measurements required for computing
+        the expectation value of a qubit hamiltonian, where co-measurable terms
+        are grouped. We're assuming the exact expectation values are provided
+        (i.e. infinite number of measurements or simulations without noise)
+        M ~ (\sum_{i} prec(H_i)) ** 2.0 / (epsilon ** 2.0)
+        where prec(H_i) is the precision (square root of the variance)
+        for each group of co-measurable terms H_{i}. It is computed as
+        prec(H_{i}) = \sum{ab} |h_{a}^{i}||h_{b}^{i}| cov(O_{a}^{i}, O_{b}^{i})
+        where h_{a}^{i} is the coefficient of the a-th operator, O_{a}^{i}, in the
+        i-th group. Covariances are assumed to be zero for a != b:
+        cov(O_{a}^{i}, O_{b}^{i}) = <O_{a}^{i} O_{b}^{i}> - <O_{a}^{i}> <O_{b}^{i}> = 0
+    Args:
+        target_operator: A QubitOperator to measure
+        expecval: An ExpectationValues object containing the expectation
+                  values of the operators and their squares. Optionally, contains
+                  values of operator products to compute covariances.
+                  If absent, covariances are assumed to be 0 and variances are
+                  assumed to be maximal, i.e. 1. It is assumed that the first expectation
+                  value corresponds to the constant term in the target operator in line 
+                  with the conventions used in the BasicEstimator
+                  NOTE: WE HAVE TO MAKE SURE THAT THE ORDER
+                  OF EXPECTATION VALUES MATCHES THE ORDER OF THE TERMS IN THE
+                  TARGET QUBIT OPERATOR, OTHERWISE THIS FUNCTION WILL NOT WORK CORRECTLY
+    Returns:
+        K2: number of measurements for epsilon = 1.0
+        nterms: number of groups of QWC terms in the target_operator
+        frame_meas: Number of optimal measurements per group 
+    """
+
+    frame_variances = None
+    if decomposition_method == 'greedy-sorted':
+        decomposition_function = lambda qubit_operator: group_comeasureable_terms_greedy(
+        qubit_operator, True) 
+    elif decomposition_method == 'greedy':
+        decomposition_function = lambda qubit_operator: group_comeasureable_terms_greedy(
+        qubit_operator, False)
+    else:
+        raise Exception(f'{decomposition_method} grouping is not implemented')
```

Comment:
We should probably raise here `ValueError` instead of `Exception`, which would make it easier to catch and handle if needed.